default/python/universe_generation/galaxy.py

```diff
@@ -1,35 +1,349 @@
+        def parents(p1, children):
+            # print "pp p1 ", p1.pos, " children a ", children
+            if p1.parent:
+                # print "pp deeper ", p1.pos, " chlidren b ", children
+                children.append(p1)
+                return parents(p1.parent, children)
+            else:
+                # print "pp done p1 ", p1.pos, " children c ", children
+                return (p1, children)
+
+        if pos not in self.dsets:
+            return []
+
+        (root, children) = parents(self.dsets[pos], [])
+        # squash the chain of children
+        for child in children:
+            child.bind_parent(root)
+        return [root.pos]
+
+    def complete_sets(self):
+        """ returns a list of list of all sets O(n)."""
+        ret = defaultdict(list)
+        for pos in self.dsets.keys():
+            ret[self.root(pos)].append(pos)
+        return ret.values()
+
+
+class Clusterer(object):
+    """
+    Computes all clusters of positions separated by more than MAX_STARLANE_LENGTH
+
+    This creates a disjoint set of all of the positions
+    Adds all of the seperations less than MAX_STARLANE_LENGTH
+    Then returns the clusters of stars when requested
+    """
+    def __init__(self, positions, adjacency_grid):
+        self.adjacency_grid = adjacency_grid
+        self.dsets = DisjointSets()
+
+        for pos in positions:
+            self._add(pos)
+
+        self.clusters = set()
+        for cluster in self.dsets.complete_sets():
+            # Decorate each cluster with its length to speed sorting
+            self.clusters.add((len(cluster), frozenset(cluster)))
+
+    def __len__(self):
+        return len(self.clusters)
+
+    def _add(self, pos):
+        """Adds pos."""
+        for neighbor in self.adjacency_grid.neighbors(pos):
+            # print "Clusterer adding ", pos,  " and ", neighbor
+            self.dsets.link(pos, neighbor)
+
+    def smallest_isolated_cluster(self):
+        """
+        Return None if all of the positions are in one cluster
+        or the smallest cluster
+        """
+        # Sort (min) and return the undecorated cluster.
+        return None if len(self.clusters) < 2 else min(self.clusters)[1]
+
+    def stitch_clusters(self, p1, p2, stitches):
+        """
+        Stitch the clusters containing ''p1'' and ''p2''
+        together with the positions in ''stitches''
+
+        This assumes that the stitching is correctly formed.
+        ''p1'' and ''p2'' are the closest positions between
+        the clusters and the positions in ''stitches'' are only
+        between ''p1'' and ''p2''
+        """
+        len_dset1 = None
+        len_dset2 = None
+        for len_dset in self.clusters:
+            if p1 in len_dset[1]:
+                len_dset1 = len_dset
+            if p2 in len_dset[1]:
+                len_dset2 = len_dset
+        assert len_dset1 and len_dset2, "p1 and p2 must be points in disjoint sets of positions"
```

Comment:
I understand what you mean, and you certainly have a point here. However, it has been project policy for as long as I can remember that as long as there is any possibility to continue execution we *never* fail hard (which is why there are so few instances where exceptions are thrown, and even where that happens, the majority gets caught and tried to handle at a higher level, at least AFAIK). Regardless if we're dealing with potential user config/input or programmer errors.

The reasoning behind that is that the game shouldn't crash on the player unless it's absolutely impossible to avoid. Even if game execution is impaired. Of course, provided these errors get logged and the logs carefully checked (especially before committing/merging things into master), to avoid the problems you mentioned.

As you can probably guess, not everyone was/is in agreement with that policy, and it has been repeatedly disputed. The reasons you cite above usually being the main argument against it.

However, that hasn't led to a change to this policy, so it's still in force. That said, I don't think that it can't be challenged again. As I said above, I do see your point, and I'm not adamant about it. But I'm not the one you need to convince, that would be @geoffthemedio, our project and programming lead. If you want to put that topic up for discussion, I suggest doing so on the forum (probably with a link to this little exchange here).

As far as our specific case here is concerned however, I still tend to prefer soft failure, because this is a part of the universe generation process that isn't vitally important to carry on. If we fail here (regardless if of an user config or programmer error), all that happens is that the max starlane length doesn't get enforced, and we might end up with a map that has some overly long starlanes. That shouldn't cause universe generation to fail entirely.

That said, I expect the risk of that assertion to ever fail very low, so we can change that at a later point, no need to hold up merging the PR because of it.